### ISSUE TITLE
"Open new tab in current directory" should not follow symlinks

### ISSUE THREAD

**Issue by Jehan** [Guake/guake#967]
I have directory symlinks in which I often works, but I want to keep seeing the non-followed path.

These kinds of stuff: $HOME/build/gimp/ and $HOME/build/ is actually a symlink to another part of the disk (which I don't save).
`pwd` or `echo $PWD` shows $HOME/build/gimp/ which is exactly what I want. But when I create a new tab, it changes to the symlinked path. I don't want this! I want the new tab to visually stay in the exact same path as the previous one. Could this be fixed?
Thanks!

**Comment by Jehan** [comment]
For info, that's what pwd calls logical vs physical working directory (-L and -P option). It supports it officially by simply checking the value of $PWD to get the logical working directory: http://git.savannah.gnu.org/cgit/coreutils.git/tree/src/pwd.c

Unfortunately I see how you get the physical WD by reading the symlink `/proc/$$/cwd` but I have no idea for the moment on how to get the logical one. My first idea was to read `/proc/{0}/environ` similarly but this is actually the calling environment, not the one inside the shell being run. I am not sure how we can get this information for the time being.

**Comment by egmontkob** [comment]
The way this works in gnome-terminal and a few other vte-based ones is:

- They're based on libvte-2.91 (gtk3) which supports the OSC 7 escape sequence that sets the terminal emulator's notion of the working directory. That is, no more digging under /proc at all, the shell tells it explicitly to the terminal emulator (as a string, potentially containing symlink components).

- Source /etc/profile.d/vte-2.91.sh at startup so that it's hooked up to the shell.

- Set the actual working directory as well as $PWD in the newly opened tabs, the latter one makes the shell keep symlink components.

**Comment by lgenzelis** [comment]
This was already [reported some years ago](https://github.com/Guake/guake/issues/78). Has there been any attempts to fix it?

**Comment by Jehan** [comment]
> Has there been any attempts to fix it?

I had a look at the code and got stuck at where to get the logical working directory from outside the shell (i.e. from the calling processus: guake), as I explained in my comment. @egmontkob gave the solution/hints on how it is possible using libvte above. From what I understood, that seems quite straightforward. Yet I did not make the time to look into it since then. Maybe someone will. Or I may end up doing so myself someday, who knows. :-)

**Comment by lgenzelis** [comment]
Maybe I am underestimating the problem because of insufficient knowledge. But this is my reasoning: the default behavior of guake is showing the current directory in the name of the tab (and by current directory I am including the possibility that the directory may be in fact a symlink to another directory). That means that guake *can* access the desired directory name (maybe not all the time, for some reason, but it definitely access the name when we `cd` into a new directory). So, why can't we store that string (the directory name) and then automatically `cd` into it when we open a new tab?

**Comment by Jehan** [comment]
> the default behavior of guake is showing the current directory in the name of the tab

Maybe there is code for this already in guake, I don't know (I don't use the feature you talk about, that's not default on my guake). And if so, that would make this even easier to implement indeed. All it probably takes is a developer willing to do so. Be my guest. :-)

Disclaimer: I am not a guake developer, only a passerby (like you, I guess?) who would also like to see this implemented and who might give it a try someday but hope someone will do so before him. :P

**Comment by lgenzelis** [comment]
> I don't use the feature you talk about, that's not default on my guake

At least I thought it was default. Anyway, you can enable that feature by going to Guake Preferences -> General and checking *Use VTE titles for tab names*.
 
> Disclaimer: I am not a guake developer, only a passerby (like you, I guess?) who would also like to see this implemented and who might give it a try someday but hope someone will do so before him. :P

Exactly the same here! :p

**Comment by egmontkob** [comment]
> Use VTE titles for tab names.

That's entirely different. As its name says, it means to actually display the title of the VTE widget, whatever that is. Now, some shell setups might put the current directory in the VTE title, but there's absolutely no guarantee for that. Some shell setups, some other apps etc. could easily show there something different. It would be a huge mistake to assume that the VTE title is the current working directory and use it for that purpose.

As said earlier, there is now a dedicated slot to store the working directory, possibly including symlink components. This, however, requires vte-2.91, so is blocked by the gtk3 upgrade.

In earlier versions, the only possibility is to dig under `/proc` to get the working directory, but that will report the canonical path without symlinks.

**Comment by lgenzelis** [comment]
>  It would be a huge mistake to assume that the VTE title is the current working directory and use it for that purpose.
I'm not familiar with VTE, but that's what I've seen with every version of Ubuntu I've used, so I jumped to that conclusion. Thanks for the clarification.

I've just checked and gnome-terminal and terminator both exhibit this same issue, and there's been users complaining about it for years. So, this is not something inherent to quake. I don't think we'll see a fix soon :(

**Comment by egmontkob** [comment]
> I've just checked and gnome-terminal and terminator both exhibit this same issue

Really? Did you check their newest vanilla versions? Both gnome-terminal and terminator _do preserve_ symlink components in the current directory, because they carry over the current directory as a string, as reported by VTE upon receiving an OSC 7 escape sequence, which is set up while sourcing `vte-2.91.sh` if it's set up correctly by your distribution.

You may easily be mislead by an earlier Terminator version (their gtk3-based release is relatively new), or Ubuntu's custom crappy patch to gnome-terminal reverting digging under /proc.

**Comment by lgenzelis** [comment]
> Did you check their newest vanilla versions?

No, I didn't. I just checked the gnome-terminal that is installed by default with ubuntu 16.04, and I installed the latest terminator **from the repositories**.  So yes, if that has been fixed recently, I probably don't have the corrected versions.

**Comment by Stibbons** [comment]
I understand the issue, eventhough if not transparently handled by VTE I don't see what to do. I hope it is fixed in gtk3
